File: src/agent/orchestrator/checkpoint.ts
/**
 * Orchestrator Checkpoint System
 *
 * Provides crash recovery for the orchestrator by persisting state at key phase
 * transitions. When the orchestrator crashes, it can resume from the last checkpoint
 * instead of restarting from the beginning.
 *
 * Design principles:
 * - Atomic writes (write to temp file, then rename)
 * - Checkpoints expire after 24 hours
 * - Git state is captured for validation on resume
 * - Compatible with existing two-phase commit recovery
 *
 * @module agent/orchestrator/checkpoint
 */

import * as FileSystem from "@effect/platform/FileSystem";
import { Effect, Option } from "effect";
import type { OrchestratorPhase } from "./types.js";

// ============================================================================
// Types
// ============================================================================

/**
 * Git state captured at checkpoint time for validation on resume.
 */
export interface CheckpointGitState {
  readonly branch: string;
  readonly headCommit: string;
  readonly isDirty: boolean;
  readonly stagedFiles: readonly string[];
}

/**
 * Verification results captured after verify phase.
 */
export interface CheckpointVerification {
  readonly typecheckPassed: boolean;
  readonly testsPassed: boolean;
  readonly verifiedAt: string;
}

/**
 * Healer invocation record for session auditing.
 */
export interface CheckpointHealerInvocation {
  readonly scenario: string;
  readonly outcome: string;
  readonly timestamp: string;
}

/**
 * Main checkpoint schema.
 * Captures orchestrator state at phase boundaries for crash recovery.
 */
export interface OrchestratorCheckpoint {
  /** Schema version for future compatibility */
  readonly version: 1;
  /** Unique session identifier */
  readonly sessionId: string;
  /** When checkpoint was written */
  readonly timestamp: string;

  // Current position in workflow
  /** Current orchestrator phase */
  readonly phase: OrchestratorPhase;
  /** When current phase started */
  readonly phaseStartedAt: string;

  // Task context
  /** ID of task being worked on */
  readonly taskId: string;
  /** Title of task for human readability */
  readonly taskTitle: string;

  // Subtask progress
  /** IDs of completed subtasks */
  readonly completedSubtaskIds: readonly string[];
  /** ID of subtask currently being executed (null if not in execute phase) */
  readonly currentSubtaskId: string | null;

  // Git state
  /** Git repository state at checkpoint */
  readonly git: CheckpointGitState;

  // Optional: Verification results (populated after verify phase)
  readonly verification?: CheckpointVerification;

  // Healer audit trail
  readonly healerInvocations: readonly CheckpointHealerInvocation[];
}

/**
 * Result of checkpoint validation.
 */
export type CheckpointValidation =
  | { readonly valid: true; readonly checkpoint: OrchestratorCheckpoint }
  | { readonly valid: false; readonly reason: string };

// ============================================================================
// Constants
// ============================================================================

/** Checkpoint file name */
export const CHECKPOINT_FILENAME = "checkpoint.json";

/** Maximum age before checkpoint is considered stale (24 hours in ms) */
export const CHECKPOINT_MAX_AGE_MS = 24 * 60 * 60 * 1000;

// ============================================================================
// Path Helpers
// ============================================================================

/**
 * Get the path to the checkpoint file.
 */
export const getCheckpointPath = (openagentsDir: string): string =>
  `${openagentsDir}/${CHECKPOINT_FILENAME}`;

/**
 * Get the path to the temporary checkpoint file (used for atomic writes).
 */
const getTempCheckpointPath = (openagentsDir: string): string =>
  `${openagentsDir}/${CHECKPOINT_FILENAME}.tmp`;

// ============================================================================
// Git State Capture
// ============================================================================

/**
 * Capture current git state for checkpoint.
 */
export const captureGitState = (
  cwd: string,
): Effect.Effect<CheckpointGitState, Error, never> =>
  Effect.tryPromise({
    try: async () => {
      const { execSync } = await import("node:child_process");

      // Get current branch
      const branch = execSync("git rev-parse --abbrev-ref HEAD", {
        cwd,
        encoding: "utf-8",
      }).trim();

      // Get HEAD commit
      const headCommit = execSync("git rev-parse HEAD", {
        cwd,
        encoding: "utf-8",
      }).trim();

      // Check for dirty state (uncommitted changes)
      let isDirty = false;
      try {
        execSync("git diff --quiet", { cwd, encoding: "utf-8" });
        execSync("git diff --staged --quiet", { cwd, encoding: "utf-8" });
      } catch {
        isDirty = true;
      }

      // Get staged files
      const stagedOutput = execSync("git diff --staged --name-only", {
        cwd,
        encoding: "utf-8",
      }).trim();
      const stagedFiles = stagedOutput ? stagedOutput.split("\n") : [];

      return {
        branch,
        headCommit,
        isDirty,
        stagedFiles,
      };
    },
    catch: (error: unknown) =>
      new Error(
        `Failed to capture git state: ${error instanceof Error ? error.message : String(error)}`,
      ),
  });

// ============================================================================
// Checkpoint Persistence
// ============================================================================

/**
 * Write checkpoint atomically (temp file + rename).
 * This ensures the checkpoint file is never in a partial/corrupt state.
 */
export const writeCheckpoint = (
  openagentsDir: string,
  checkpoint: OrchestratorCheckpoint,
): Effect.Effect<void, Error, FileSystem.FileSystem> =>
  Effect.gen(function* () {
    const fs = yield* FileSystem.FileSystem;
    const checkpointPath = getCheckpointPath(openagentsDir);
    const tempPath = getTempCheckpointPath(openagentsDir);

    // Write to temp file
    yield* fs.writeFileString(tempPath, JSON.stringify(checkpoint, null, 2)).pipe(
      Effect.mapError(
        (e) => new Error(`Failed to write temp checkpoint: ${e.message}`),
      ),
    );

    // Atomic rename
    yield* fs.rename(tempPath, checkpointPath).pipe(
      Effect.mapError(
        (e) => new Error(`Failed to rename checkpoint: ${e.message}`),
      ),
    );
  });

/**
 * Read existing checkpoint file.
 * Returns Option.none() if checkpoint doesn't exist.
 */
export const readCheckpoint = (
  openagentsDir: string,
): Effect.Effect<Option.Option<OrchestratorCheckpoint>, Error, FileSystem.FileSystem> =>
  Effect.gen(function* () {
    const fs = yield* FileSystem.FileSystem;
    const checkpointPath = getCheckpointPath(openagentsDir);

    const exists = yield* fs.exists(checkpointPath);
    if (!exists) {
      return Option.none();
    }

    const content = yield* fs.readFileString(checkpointPath).pipe(
      Effect.mapError(
        (e) => new Error(`Failed to read checkpoint: ${e.message}`),
      ),
    );

    try {
      const checkpoint = JSON.parse(content) as OrchestratorCheckpoint;
      return Option.some(checkpoint);
    } catch (e) {
      return Option.none(); // Corrupted checkpoint, treat as non-existent
    }
  });

/**
 * Clear checkpoint file (on successful completion).
 */
export const clearCheckpoint = (
  openagentsDir: string,
): Effect.Effect<void, never, FileSystem.FileSystem> =>
  Effect.gen(function* () {
    const fs = yield* FileSystem.FileSystem;
    const checkpointPath = getCheckpointPath(openagentsDir);

    yield* fs.remove(checkpointPath).pipe(Effect.ignore);

    // Also clean up temp file if it exists
    const tempPath = getTempCheckpointPath(openagentsDir);
    yield* fs.remove(tempPath).pipe(Effect.ignore);
  });

// ============================================================================
// Checkpoint Validation
// ============================================================================

/**
 * Validate a checkpoint for resumption.
 *
 * Checks:
 * - Checkpoint is not stale (>24 hours old)
 * - Git branch matches current branch
 * - Version is compatible
 */
export const validateCheckpoint = (
  checkpoint: OrchestratorCheckpoint,
  currentGitState: CheckpointGitState,
): CheckpointValidation => {
  // Check version
  if (checkpoint.version !== 1) {
    return {
      valid: false,
      reason: `Unsupported checkpoint version: ${checkpoint.version}`,
    };
  }

  // Check age
  const age = Date.now() - new Date(checkpoint.timestamp).getTime();
  if (age > CHECKPOINT_MAX_AGE_MS) {
    const hoursOld = Math.round(age / (60 * 60 * 1000));
    return {
      valid: false,
      reason: `Checkpoint is stale (${hoursOld} hours old, max 24 hours)`,
    };
  }

  // Check branch matches
  if (checkpoint.git.branch !== currentGitState.branch) {
    return {
      valid: false,
      reason: `Branch mismatch: checkpoint on '${checkpoint.git.branch}', now on '${currentGitState.branch}'`,
    };
  }

  return { valid: true, checkpoint };
};

/**
 * Check for and validate existing checkpoint.
 * Returns None if no valid checkpoint exists.
 */
export const maybeResumeCheckpoint = (
  openagentsDir: string,
  cwd: string,
): Effect.Effect<Option.Option<OrchestratorCheckpoint>, Error, FileSystem.FileSystem> =>
  Effect.gen(function* () {
    // Try to read existing checkpoint
    const maybeCheckpoint = yield* readCheckpoint(openagentsDir);

    if (Option.isNone(maybeCheckpoint)) {
      return Option.none();
    }

    const checkpoint = maybeCheckpoint.value;

    // Capture current git state
    const currentGitState = yield* captureGitState(cwd);

    // Validate checkpoint
    const validation = validateCheckpoint(checkpoint, currentGitState);

    if (!validation.valid) {
      // Invalid checkpoint - clear it and return none
      yield* clearCheckpoint(openagentsDir);
      return Option.none();
    }

    return Option.some(checkpoint);
  });

// ============================================================================
// Checkpoint Creation Helpers
// ============================================================================

/**
 * Create a new checkpoint with the given parameters.
 */
export const createCheckpoint = (params: {
  sessionId: string;
  phase: OrchestratorPhase;
  taskId: string;
  taskTitle: string;
  completedSubtaskIds: readonly string[];
  currentSubtaskId: string | null;
  git: CheckpointGitState;
  verification?: CheckpointVerification;
  healerInvocations?: readonly CheckpointHealerInvocation[];
}): OrchestratorCheckpoint => {
  const now = new Date().toISOString();
  const base = {
    version: 1 as const,
    sessionId: params.sessionId,
    timestamp: now,
    phase: params.phase,
    phaseStartedAt: now,
    taskId: params.taskId,
    taskTitle: params.taskTitle,
    completedSubtaskIds: params.completedSubtaskIds,
    currentSubtaskId: params.currentSubtaskId,
    git: params.git,
    healerInvocations: params.healerInvocations ?? [],
  };

  // Only include verification if explicitly provided (exactOptionalPropertyTypes)
  if (params.verification !== undefined) {
    return { ...base, verification: params.verification };
  }
  return base;
};

/**
 * Update an existing checkpoint with new phase.
 */
export const updateCheckpointPhase = (
  checkpoint: OrchestratorCheckpoint,
  phase: OrchestratorPhase,
  updates?: Partial<
    Pick<
      OrchestratorCheckpoint,
      "completedSubtaskIds" | "currentSubtaskId" | "verification" | "git"
    >
  >,
): OrchestratorCheckpoint => {
  const now = new Date().toISOString();
  return {
    ...checkpoint,
    timestamp: now,
    phase,
    phaseStartedAt: now,
    ...(updates?.completedSubtaskIds !== undefined && {
      completedSubtaskIds: updates.completedSubtaskIds,
    }),
    ...(updates?.currentSubtaskId !== undefined && {
      currentSubtaskId: updates.currentSubtaskId,
    }),
    ...(updates?.verification !== undefined && {
      verification: updates.verification,
    }),
    ...(updates?.git !== undefined && { git: updates.git }),
  };
};

/**
 * Add a healer invocation to the checkpoint.
 */
export const addHealerInvocation = (
  checkpoint: OrchestratorCheckpoint,
  invocation: CheckpointHealerInvocation,
): OrchestratorCheckpoint => ({
  ...checkpoint,
  timestamp: new Date().toISOString(),
  healerInvocations: [...checkpoint.healerInvocations, invocation],
});

File: crates/agent_openagents_backup/src/checkpoint.rs
//! Orchestrator Checkpoint System
//!
//! Provides crash recovery for the orchestrator by persisting state at key phase
//! transitions. When the orchestrator crashes, it can resume from the last checkpoint
//! instead of restarting from the beginning.
//!
//! Design principles:
//! - Atomic writes (write to temp file, then rename)
//! - Checkpoints expire after 24 hours
//! - Git state is captured for validation on resume

use crate::error::AgentResult;
use crate::types::OrchestratorPhase;
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::fs;
use std::path::Path;

/// Checkpoint file name
pub const CHECKPOINT_FILENAME: &str = "checkpoint.json";

/// Maximum age before checkpoint is considered stale (24 hours in ms)
pub const CHECKPOINT_MAX_AGE_MS: i64 = 24 * 60 * 60 * 1000;

/// Git state captured at checkpoint time for validation on resume
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CheckpointGitState {
    pub branch: String,
    pub head_commit: String,
    pub is_dirty: bool,
    pub staged_files: Vec<String>,
}

/// Verification results captured after verify phase
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CheckpointVerification {
    pub typecheck_passed: bool,
    pub tests_passed: bool,
    pub verified_at: String,
}

/// Healer invocation record for session auditing
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CheckpointHealerInvocation {
    pub scenario: String,
    pub outcome: String,
    pub timestamp: String,
}

/// Main checkpoint schema.
/// Captures orchestrator state at phase boundaries for crash recovery.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OrchestratorCheckpoint {
    /// Schema version for future compatibility
    pub version: u32,
    /// Unique session identifier
    pub session_id: String,
    /// When checkpoint was written
    pub timestamp: String,
    /// Current orchestrator phase
    pub phase: OrchestratorPhase,
    /// When current phase started
    pub phase_started_at: String,
    /// ID of task being worked on
    pub task_id: String,
    /// Title of task for human readability
    pub task_title: String,
    /// IDs of completed subtasks
    pub completed_subtask_ids: Vec<String>,
    /// ID of subtask currently being executed
    #[serde(skip_serializing_if = "Option::is_none")]
    pub current_subtask_id: Option<String>,
    /// Git repository state at checkpoint
    pub git: CheckpointGitState,
    /// Verification results (populated after verify phase)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub verification: Option<CheckpointVerification>,
    /// Healer audit trail
    pub healer_invocations: Vec<CheckpointHealerInvocation>,
}

/// Result of checkpoint validation
#[derive(Debug, Clone)]
pub enum CheckpointValidation {
    Valid(OrchestratorCheckpoint),
    Invalid { reason: String },
}

impl CheckpointValidation {
    pub fn is_valid(&self) -> bool {
        matches!(self, Self::Valid(_))
    }
}

/// Get the path to the checkpoint file
pub fn get_checkpoint_path(openagents_dir: &str) -> String {
    format!("{}/{}", openagents_dir, CHECKPOINT_FILENAME)
}

/// Get the path to the temporary checkpoint file (used for atomic writes)
fn get_temp_checkpoint_path(openagents_dir: &str) -> String {
    format!("{}/{}.tmp", openagents_dir, CHECKPOINT_FILENAME)
}

/// Write checkpoint atomically (temp file + rename)
pub fn write_checkpoint(openagents_dir: &str, checkpoint: &OrchestratorCheckpoint) -> AgentResult<()> {
    let checkpoint_path = get_checkpoint_path(openagents_dir);
    let temp_path = get_temp_checkpoint_path(openagents_dir);

    // Ensure directory exists
    if let Some(parent) = Path::new(&checkpoint_path).parent() {
        fs::create_dir_all(parent)?;
    }

    // Write to temp file
    let content = serde_json::to_string_pretty(checkpoint)?;
    fs::write(&temp_path, &content)?;

    // Atomic rename
    fs::rename(&temp_path, &checkpoint_path)?;

    Ok(())
}

/// Read existing checkpoint file
pub fn read_checkpoint(openagents_dir: &str) -> Option<OrchestratorCheckpoint> {
    let checkpoint_path = get_checkpoint_path(openagents_dir);
    let path = Path::new(&checkpoint_path);

    if !path.exists() {
        return None;
    }

    let content = fs::read_to_string(path).ok()?;
    serde_json::from_str(&content).ok()
}

/// Clear checkpoint file (on successful completion)
pub fn clear_checkpoint(openagents_dir: &str) {
    let checkpoint_path = get_checkpoint_path(openagents_dir);
    let temp_path = get_temp_checkpoint_path(openagents_dir);

    let _ = fs::remove_file(&checkpoint_path);
    let _ = fs::remove_file(&temp_path);
}

/// Validate a checkpoint for resumption
pub fn validate_checkpoint(
    checkpoint: &OrchestratorCheckpoint,
    current_git_state: &CheckpointGitState,
) -> CheckpointValidation {
    // Check version
    if checkpoint.version != 1 {
        return CheckpointValidation::Invalid {
            reason: format!("Unsupported checkpoint version: {}", checkpoint.version),
        };
    }

    // Check age
    if let Ok(checkpoint_time) = DateTime::parse_from_rfc3339(&checkpoint.timestamp) {
        let age_ms = Utc::now()
            .signed_duration_since(checkpoint_time)
            .num_milliseconds();
        if age_ms > CHECKPOINT_MAX_AGE_MS {
            let hours_old = age_ms / (60 * 60 * 1000);
            return CheckpointValidation::Invalid {
                reason: format!("Checkpoint is stale ({} hours old, max 24 hours)", hours_old),
            };
        }
    }

    // Check branch matches
    if checkpoint.git.branch != current_git_state.branch {
        return CheckpointValidation::Invalid {
            reason: format!(
                "Branch mismatch: checkpoint on '{}', now on '{}'",
                checkpoint.git.branch, current_git_state.branch
            ),
        };
    }

    CheckpointValidation::Valid(checkpoint.clone())
}

/// Create a new checkpoint
pub fn create_checkpoint(
    session_id: &str,
    phase: OrchestratorPhase,
    task_id: &str,
    task_title: &str,
    completed_subtask_ids: Vec<String>,
    current_subtask_id: Option<String>,
    git: CheckpointGitState,
    verification: Option<CheckpointVerification>,
    healer_invocations: Vec<CheckpointHealerInvocation>,
) -> OrchestratorCheckpoint {
    let now = Utc::now().to_rfc3339();
    OrchestratorCheckpoint {
        version: 1,
        session_id: session_id.to_string(),
        timestamp: now.clone(),
        phase,
        phase_started_at: now,
        task_id: task_id.to_string(),
        task_title: task_title.to_string(),
        completed_subtask_ids,
        current_subtask_id,
        git,
        verification,
        healer_invocations,
    }
}

/// Update an existing checkpoint with new phase
pub fn update_checkpoint_phase(
    checkpoint: &OrchestratorCheckpoint,
    phase: OrchestratorPhase,
    updates: Option<CheckpointUpdates>,
) -> OrchestratorCheckpoint {
    let now = Utc::now().to_rfc3339();
    let updates = updates.unwrap_or_default();

    OrchestratorCheckpoint {
        version: checkpoint.version,
        session_id: checkpoint.session_id.clone(),
        timestamp: now.clone(),
        phase,
        phase_started_at: now,
        task_id: checkpoint.task_id.clone(),
        task_title: checkpoint.task_title.clone(),
        completed_subtask_ids: updates
            .completed_subtask_ids
            .unwrap_or_else(|| checkpoint.completed_subtask_ids.clone()),
        current_subtask_id: updates
            .current_subtask_id
            .or_else(|| checkpoint.current_subtask_id.clone()),
        git: updates.git.unwrap_or_else(|| checkpoint.git.clone()),
        verification: updates.verification.or_else(|| checkpoint.verification.clone()),
        healer_invocations: checkpoint.healer_invocations.clone(),
    }
}

/// Updates for checkpoint phase change
#[derive(Debug, Clone, Default)]
pub struct CheckpointUpdates {
    pub completed_subtask_ids: Option<Vec<String>>,
    pub current_subtask_id: Option<String>,
    pub git: Option<CheckpointGitState>,
    pub verification: Option<CheckpointVerification>,
}

/// Add a healer invocation to the checkpoint
pub fn add_healer_invocation(
    checkpoint: &OrchestratorCheckpoint,
    invocation: CheckpointHealerInvocation,
) -> OrchestratorCheckpoint {
    let mut new_checkpoint = checkpoint.clone();
    new_checkpoint.timestamp = Utc::now().to_rfc3339();
    new_checkpoint.healer_invocations.push(invocation);
    new_checkpoint
}

#[cfg(test)]
mod tests {
    use super::*;

    fn make_git_state() -> CheckpointGitState {
        CheckpointGitState {
            branch: "main".to_string(),
            head_commit: "abc123".to_string(),
            is_dirty: false,
            staged_files: vec![],
        }
    }

    #[test]
    fn test_get_checkpoint_path() {
        assert_eq!(
            get_checkpoint_path("/home/user/.openagents"),
            "/home/user/.openagents/checkpoint.json"
        );
    }

    #[test]
    fn test_create_checkpoint() {
        let git = make_git_state();
        let checkpoint = create_checkpoint(
            "session-1",
            OrchestratorPhase::Decomposing,
            "task-1",
            "Fix bug",
            vec!["sub-1".to_string()],
            Some("sub-2".to_string()),
            git,
            None,
            vec![],
        );

        assert_eq!(checkpoint.version, 1);
        assert_eq!(checkpoint.session_id, "session-1");
        assert_eq!(checkpoint.phase, OrchestratorPhase::Decomposing);
        assert_eq!(checkpoint.task_id, "task-1");
        assert_eq!(checkpoint.completed_subtask_ids, vec!["sub-1"]);
        assert_eq!(checkpoint.current_subtask_id, Some("sub-2".to_string()));
    }

    #[test]
    fn test_validate_checkpoint_version() {
        let git = make_git_state();
        let mut checkpoint = create_checkpoint(
            "session-1",
            OrchestratorPhase::Decomposing,
            "task-1",
            "Fix bug",
            vec![],
            None,
            git.clone(),
            None,
            vec![],
        );
        checkpoint.version = 99;

        let result = validate_checkpoint(&checkpoint, &git);
        assert!(!result.is_valid());
    }

    #[test]
    fn test_validate_checkpoint_branch_mismatch() {
        let git = make_git_state();
        let checkpoint = create_checkpoint(
            "session-1",
            OrchestratorPhase::Decomposing,
            "task-1",
            "Fix bug",
            vec![],
            None,
            git.clone(),
            None,
            vec![],
        );

        let different_branch = CheckpointGitState {
            branch: "feature-branch".to_string(),
            ..git
        };

        let result = validate_checkpoint(&checkpoint, &different_branch);
        assert!(!result.is_valid());
    }

    #[test]
    fn test_update_checkpoint_phase() {
        let git = make_git_state();
        let checkpoint = create_checkpoint(
            "session-1",
            OrchestratorPhase::Decomposing,
            "task-1",
            "Fix bug",
            vec![],
            None,
            git,
            None,
            vec![],
        );

        let updates = CheckpointUpdates {
            completed_subtask_ids: Some(vec!["sub-1".to_string()]),
            ..Default::default()
        };

        let updated = update_checkpoint_phase(&checkpoint, OrchestratorPhase::Verifying, Some(updates));
        assert_eq!(updated.phase, OrchestratorPhase::Verifying);
        assert_eq!(updated.completed_subtask_ids, vec!["sub-1"]);
    }

    #[test]
    fn test_add_healer_invocation() {
        let git = make_git_state();
        let checkpoint = create_checkpoint(
            "session-1",
            OrchestratorPhase::Decomposing,
            "task-1",
            "Fix bug",
            vec![],
            None,
            git,
            None,
            vec![],
        );

        let invocation = CheckpointHealerInvocation {
            scenario: "typecheck_failed".to_string(),
            outcome: "fixed".to_string(),
            timestamp: Utc::now().to_rfc3339(),
        };

        let updated = add_healer_invocation(&checkpoint, invocation);
        assert_eq!(updated.healer_invocations.len(), 1);
        assert_eq!(updated.healer_invocations[0].scenario, "typecheck_failed");
    }
}
